### PR TITLE
Periodically saving checkpoint after processing given number of batches

### PR DESCRIPTION
**Pull request by csukuangfj**
Fixes https://github.com/lhotse-speech/lhotse/issues/627

### CODE REVIEW

File: icefall/checkpoint.py [hunk 1]
```diff
@@ -151,3 +154,99 @@ def average_checkpoints(
             avg[k] //= n
 
     return avg
+
+
+def save_checkpoint_with_global_batch_idx(
```
Comment (pzelasko) [inline icefall/checkpoint.py:159]: I think that you should also pass `sampler` here and save its state dict.

**Comment by csukuangfj** [comment]
# Usage

## Training
(1) 
```
--save-every-n 8000
--keep-last-k 5
```
means to save checkpoints whenever `params.batch_idx_train % 8000 == 0` and it keeps only the last 5 saved checkpoints on disk.

(2) Resume from some checkpoints
```
--start-batch 1000
```
means to resume the training by loading `checkpoint-1000.pt`

# Decoding
```
--avg-last-n 5
```
means to average the last 5 checkpoints (i.e., the last 5 checkpoint-xxx.pt) for decoding.

----

This PR is backward compatible and is ready for review.

File: icefall/checkpoint.py (hunk 1 above)
Comment (csukuangfj) [reply icefall/checkpoint.py:164]: Yes, you are right.

File: egs/librispeech/ASR/pruned_transducer_stateless/train.py [hunk 2]
```diff
@@ -186,6 +197,30 @@ def get_parser():
...
+        has the form: f'exp-dir/checkpoint-{params.batch_idx_train}.pt'
+        Note: It also saves checkpoint to `exp-dir/epoch-xxx.pt` at the
+        end of each epoch where `xxx` is the epoch number counting from 0.
+        """,
+    )
+
+    parser.add_argument(
+        "--keep-last-k",
+        type=int,
+        default=10,
```
Comment (danpovey) [inline egs/librispeech/ASR/pruned_transducer_stateless/train.py:216]: I suspect we might want to average more than 10, typically (depending how frequently they are saved).

**Comment by danpovey** [comment]
Very nice!

File: egs/librispeech/ASR/pruned_transducer_stateless/train.py (hunk 2 above)
Comment (csukuangfj) [reply egs/librispeech/ASR/pruned_transducer_stateless/train.py:216]: 10 is just the default value and can be provided by users from the command line.
What default value do you recommend here?

Comment (danpovey) [reply egs/librispeech/ASR/pruned_transducer_stateless/train.py:216]: I'm thinking maybe 20.

**Comment by danpovey** [comment]
OK with me to merge if you think it most likely won't break existing scripts..

**Comment by danpovey** [comment]
```
tail pruned_transducer_stateless2/rework2f_sequential_lr10/log/log-train-2022-03-23-11-41-53
2022-03-23 11:41:53,668 INFO [train.py:702] Training started
2022-03-23 11:41:53,689 INFO [train.py:712] Device: cuda:0
2022-03-23 11:41:53,695 INFO [train.py:721] {'best_train_loss': inf, 'best_valid_loss': inf, 'best_train_epoch': -1, 'best_valid_epoch': -1, 'batch_idx_train': 0, 'log_interval': 50, 'reset_interval': 200, 'valid_in\
terval': 1600, 'feature_dim': 80, 'subsampling_factor': 4, 'attention_dim': 512, 'nhead': 8, 'dim_feedforward': 2048, 'num_encoder_layers': 12, 'vgg_frontend': False, 'embedding_dim': 512, 'warm_step': 30000, 'model\
_warm_step': 3000, 'env_info': {'k2-version': '1.14', 'k2-build-type': 'Release', 'k2-with-cuda': True, 'k2-git-sha1': '6833270cb228aba7bf9681fccd41e2b52f7d984c', 'k2-git-date': 'Wed Mar 16 11:16:05 2022', 'lhotse-v\
ersion': '1.0.0.dev+git.e88c2a2.clean', 'torch-cuda-available': True, 'torch-cuda-version': '10.1', 'python-version': '3.8', 'icefall-git-branch': 'rework2f_sequential', 'icefall-git-sha1': 'c246754-clean', 'icefall\
-git-date': 'Tue Mar 22 21:44:23 2022', 'icefall-path': '/ceph-dan/icefall', 'k2-path': '/ceph-dan/.local/lib/python3.8/site-packages/k2-1.14.dev20220317+cuda10.1.torch1.7.1-py3.8-linux-x86_64.egg/k2/__init__.py', '\
lhotse-path': '/ceph-dan/.local/lib/python3.8/site-packages/lhotse-1.0.0.dev0+git.e88c2a2.clean-py3.8.egg/lhotse/__init__.py', 'hostname': 'de-74279-k2-train-6-0309102839-6644d5fdfd-gz27x', 'IP address': '10.177.28.\
71'}, 'world_size': 1, 'master_port': 12354, 'tensorboard': True, 'num_epochs': 20, 'start_epoch': 14, 'start_batch': 0, 'exp_dir': PosixPath('pruned_transducer_stateless2/rework2f_sequential_lr10'), 'bpe_model': 'd\
ata/lang_bpe_500/bpe.model', 'lr_factor': 10.0, 'context_size': 2, 'prune_range': 5, 'lm_scale': 0.25, 'am_scale': 0.0, 'simple_loss_scale': 0.5, 'seed': 42, 'print_diagnostics': False, 'save_every_n': 8000, 'keep_l\
ast_k': 20, 'full_libri': False, 'manifest_dir': PosixPath('data/fbank'), 'max_duration': 300, 'bucketing_sampler': True, 'num_buckets': 30, 'concatenate_cuts': False, 'duration_factor': 1.0, 'gap': 1.0, 'on_the_fly\
_feats': False, 'shuffle': True, 'return_cuts': True, 'num_workers': 2, 'enable_spec_aug': True, 'spec_aug_time_warp_factor': 80, 'enable_musan': True, 'blank_id': 0, 'vocab_size': 500}
2022-03-23 11:41:53,695 INFO [train.py:723] About to create model
2022-03-23 11:41:58,419 INFO [train.py:727] Number of model parameters: 84440308
2022-03-23 11:41:58,424 INFO [checkpoint.py:100] Loading checkpoint from pruned_transducer_stateless2/rework2f_sequential_lr10/epoch-13.pt
de-74279-k2-train-6-0309102839-6644d5fdfd-gz27x:ASR: 
de-74279-k2-train-6-0309102839-6644d5fdfd-gz27x:ASR: Traceback (most recent call last):
  File "./pruned_transducer_stateless2/train.py", line 894, in <module>
    main()
  File "./pruned_transducer_stateless2/train.py", line 887, in main
    run(rank=0, world_size=1, args=args)
  File "./pruned_transducer_stateless2/train.py", line 729, in run
    checkpoints = load_checkpoint_if_available(params=params, model=model)
  File "./pruned_transducer_stateless2/train.py", line 404, in load_checkpoint_if_available
    params[k] = saved_params[k]
KeyError: 'cur_batch_idx'

[1]+  Exit 1                  python3 ./pruned_transducer_stateless2/train.py --exp-dir=pruned_transducer_stateless2/rework2f_sequential_lr10 --world-size 1 --num-epochs 20 --start-epoch 14 --full-libri 0 --max-dura\
tion 300 --lr-factor 10.0 --print-diagnostics=False
```
.. any thoughts?

**Pull request merged**